### PR TITLE
Add N4

### PR DESCRIPTION
**Pull request by arnaudbore**
So far so good, didn't test anything yet and feel like we quickly need to come up with a way to handle test data.

### CODE REVIEW

**Comment by AlexVCaron** [comment]
Can you remove the auto-generated TODOs ? I'll do a first pass afterwards, but at first glance the module looks good

**Comment by arnaudbore** [comment]
> Can you remove the auto-generated TODOs ? I'll do a first pass afterwards, but at first glance the module looks good

Done

File: modules/nf-scil/preproc/n4/main.nf [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'scil.usherbrooke.ca/containers/scilus_1.5.0.sif':
+        'scilus/scilus:1.5.0' }"
+
+    input:
+    tuple val(meta), path(dwi), path(b0), path(b0_mask)
+
+    output:
+    tuple val(meta), path("*dwi_n4.nii.gz")            , emit: dwi
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/main.nf:13]: Align the emit: clauses for both outputs

File: modules/nf-scil/preproc/n4/meta.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/yaml-schema.json
+name: "preproc_n4"
+description: Bias field correction using N4
+keywords:
+  - "ANTs"
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/meta.yml:6]: Move in the tools section if you want to describe ANTs and scilpy. Else here just put some keywords.

File: modules/nf-scil/preproc/n4/meta.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    description: ANTs computes high-dimensional mappings to capture the statistics of brain structure and function.
+
+  - "scilpy"
+    description: Scilpy is the main library supporting research and development at the Sherbrooke Connectivity Imaging Lab
+tools:
+  - "preproc":
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/meta.yml:12]: Remove "preproc". It is autogenerated based on the category we specify, but we don't want it

File: modules/nf-scil/preproc/n4/meta.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'test', single_end:false ]`
+  
+  - dwi:
+      type: file
+      description: Nifti image file to correct
+      pattern: "*.{nii,nii.gz}"
+  - b0:
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/meta.yml:31]: missing newline

File: modules/nf-scil/preproc/n4/meta.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'test', single_end:false ]`
+  
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - dwi:
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/meta.yml:52]: missing newline

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
Minor comments. Can you add dummy lightweight test data so the whole test suite runs for the POC ? I would like to have a working example next thursday with all the modules we made.

We will address a new test data framework then. I intend to play with DVC a little bit, see what we can get from it.

**Comment by arnaudbore** [comment]
All the comments have been fixed but end up with an error I could not resolve...

File: tests/modules/nf-scil/preproc/n4/main.nf [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+include { PREPROC_N4 } from '../../../../../modules/nf-scil/preproc/n4/main.nf'
+
+workflow test_preproc_n4 {
+    
+    input = [
+        [ id:'test', single_end:false ], // meta map
+        file(params.test_data['preproc']['n4']['dwi'], checkIfExists: true)
+        file(params.test_data['preproc']['n4']['b0'], checkIfExists: true)
+        file(params.test_data['preproc']['n4']['b0_mask'], checkIfExists: true)
```
Comment (AlexVCaron) [inline tests/modules/nf-scil/preproc/n4/main.nf:13]: Bad key : 'dwi', should be 'image'

File: tests/modules/nf-scil/preproc/n4/main.nf [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { PREPROC_N4 } from '../../../../../modules/nf-scil/preproc/n4/main.nf'
+
+workflow test_preproc_n4 {
+    
+    input = [
```
Comment (AlexVCaron) [inline tests/modules/nf-scil/preproc/n4/main.nf:9]: Check the commas between the list's items

File: tests/config/test_data.config [hunk 8]
```diff
@@ -2,12 +2,20 @@
...
             "cropvolume" {
-                image = "${params.test_data_base}/main/.test_data/scilpy/crop_volume_image.nii.gz"
-                mask = "${params.test_data_base}/main/.test_data/scilpy/crop_volume_mask.nii.gz"
+                image = "${params.test_data_base}/main/.test_data/scilpy/dwi.nii.gz"
+                mask = "${params.test_data_base}/main/.test_data/scilpy/mask.nii.gz"
+            }
+        }
+        "preproc" {
+            "n4" {
+                image = "${params.test_data_base_n4}/main/.test_data/preproc/dwi.nii.gz"
```
Comment (AlexVCaron) [inline tests/config/test_data.config:16]: Replace `main` with the name of your source branch

File: tests/modules/nf-scil/preproc/n4/test.yml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+- name: "preproc n4"
+  command: nextflow run ./tests/modules/nf-scil/preproc/n4 -entry test_preproc_n4 -c ./tests/config/nextflow.config
+  tags:
+    - "preproc"
+    - "preproc/n4"
+  files:
+    - path: "output/preproc/test_dwi_n4.nii.gz"
```
Comment (AlexVCaron) [inline tests/modules/nf-scil/preproc/n4/test.yml:7]: Did you generate the md5sum yourself ? Nextflow does not like it on my end when running the tests. If yes, just remove it, I will regen all md5sum anyway when we have the true test infra.

File: modules/nf-scil/preproc/n4/main.nf [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        -o [${prefix}__b0_n4.nii.gz, bias_field_b0.nii.gz]\
+        -c [300x150x75x50, 1e-6] -v 1
+
+    scil_apply_bias_field_on_dwi.py $dwi bias_field_b0.nii.gz\
+        ${prefix}__dwi_n4.nii.gz --mask $b0_mask -f
+
+    cat <<-END_VERSIONS > versions.yml
+        "${task.process}":
+            scilpy: 1.6.0
+            N4BiasFieldCorrection: \$(N4BiasFieldCorrection --version 2>&1 | sed -n 's/ANTs Version: \\([0-9.]\\+\\)/\\1/p')
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/main.nf:37]:
```suggestion
            N4BiasFieldCorrection: \$(N4BiasFieldCorrection --version 2>&1 | sed -n 's/ANTs Version: v\\([0-9.]\\+\\)/\\1/p')
```

File: modules/nf-scil/preproc/n4/main.nf [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    N4BiasFieldCorrection.py -h
+    scil_apply_bias_field_on_dwi -h
+
+    touch ${prefix}_dwi_n4.nii.gz
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        scilpy: 1.6.0
+        N4BiasFieldCorrection: \$(N4BiasFieldCorrection --version 2>&1 | sed -n 's/ANTs Version: \\([0-9.]\\+\\)/\\1/p')
```
Comment (AlexVCaron) [inline modules/nf-scil/preproc/n4/main.nf:53]:
```suggestion
            N4BiasFieldCorrection: \$(N4BiasFieldCorrection --version 2>&1 | sed -n 's/ANTs Version: v\\([0-9.]\\+\\)/\\1/p')
```

File: tests/config/test_data.config [hunk 12]
```diff
@@ -2,12 +2,20 @@
...
             "cropvolume" {
-                image = "${params.test_data_base}/main/.test_data/scilpy/crop_volume_image.nii.gz"
-                mask = "${params.test_data_base}/main/.test_data/scilpy/crop_volume_mask.nii.gz"
+                image = "${params.test_data_base}/main/.test_data/scilpy/dwi.nii.gz"
+                mask = "${params.test_data_base}/main/.test_data/scilpy/mask.nii.gz"
+            }
+        }
+        "preproc" {
+            "n4" {
+                dwi = "${params.test_data_base_n4}/n4_module/.test_data/preproc/dwi.nii.gz"
```
Comment (AlexVCaron) [inline tests/config/test_data.config:16]: You can switch back those endpoints to target the `main` branch on `scilus` nf-scil

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
Last fix to make the version extraction for N4 work. Don't forget to set the test_data endpoint back. Then we can merge !

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
Last fix to make the version extraction for N4 work. Don't forget to set the test_data endpoint back. Then we can merge !

**Review by AlexVCaron — APPROVED** [review]

**Pull request merged**